edge/pkg/edged/edged.go

```diff
@@ -473,6 +540,42 @@ func newEdged() (*edged, error) {
 				conf.remoteImageEndpoint = conf.remoteRuntimeEndpoint
 			}
 		}
+		//create and start the docker shim running as a grpc server
+		if conf.remoteRuntimeEndpoint == DockerShimEndpoint || conf.remoteRuntimeEndpoint == DockerShimEndpointDeprecated {
+			streamingConfig := &streaming.Config{}
+			DockerClientConfig := &dockershim.ClientConfig{
+				DockerEndpoint:    DockerEndpoint,
+				EnableSleep:       true,
+				WithTraceDisabled: true,
+			}
+
+			pluginConfigs := dockershim.NetworkPluginSettings{
+				HairpinMode:       kubeletinternalconfig.HairpinMode(HairpinMod),
+				NonMasqueradeCIDR: NonMasqueradeCIDR,
+				PluginName:        pluginName,
+				PluginBinDir:      pluginBinDir,
+				PluginConfDir:     pluginConfDir,
+				MTU:               mtu,
+			}
+
+			redirectContainerStream := redirectContainerStream
+			cgroupDriver := ed.cgroupDriver
+
+			ds, err := dockershim.NewDockerService(DockerClientConfig, conf.PodSandboxImage, streamingConfig,
+				&pluginConfigs, croupName, cgroupDriver, DockershimRootDir, redirectContainerStream)
+
+			if err != nil {
+				return nil, err
+			}
+
+			server := dockerremote.NewDockerServer(conf.remoteRuntimeEndpoint, ds)
+			if err := server.Start(); err != nil {
+				return nil, err
+			}
+
+		}
+		ed.clusterDNS = convertStrToIP(conf.clusterDNS)
+		ed.dnsConfigurer = kubedns.NewConfigurer(recorder, nodeRef, ed.nodeIP, ed.clusterDNS, conf.clusterDomain, ResolvConfDefault) //TODO
```

Comment:
what will happen if there is no kubedns available?

Reply:
How to understand kubedns，do you mean the clusterDNS ?  The "clusterDNS" works with the dnsPolicy configuration at the deployment file